src/Routing/RouteCollection.php

```diff
@@ -123,21 +135,39 @@ public function add(Route $route, array $options = []): void
     public function parse(string $url, string $method = ''): array
     {
         $decoded = urldecode($url);
+        if ($decoded !== '/') {
+            $decoded = rtrim($decoded, '/');
+        }
+
+        $queryParameters = [];
+        if (strpos($url, '?') !== false) {
+            [$url, $qs] = explode('?', $url, 2);
+            parse_str($qs, $queryParameters);
+        }
+
+        if (isset($this->staticPaths[$decoded])) {
+            foreach ($this->staticPaths[$decoded] as $route) {
+                $r = $route->parse($url, $method);
+                if ($r === null) {
+                    continue;
+                }
+
+                if ($queryParameters) {
+                    $r['?'] = $queryParameters;
+                }
+
+                return $r;
+            }
+        }
 
         // Sort path segments matching longest paths first.
-        krsort($this->_paths);
+        krsort($this->dynamicPaths);
 
-        foreach ($this->_paths as $path => $routes) {
+        foreach ($this->dynamicPaths as $path => $routes) {
             if (strpos($decoded, $path) !== 0) {
                 continue;
             }
 
-            $queryParameters = [];
-            if (strpos($url, '?') !== false) {
-                [$url, $qs] = explode('?', $url, 2);
-                parse_str($qs, $queryParameters);
-            }
```

Comment:
Oh wow, this didn't need to be done on each loop iteration :facepalm:

Reply:
Yeah, this particular change can be done for 4.x too.